presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,285 @@
+ */
+public class KHyperLogLog
+{
+    public static final int DEFAULT_HLL_BUCKETS = 256;
+    public static final int DEFAULT_MAX_SIZE = 4096;
+    private static final byte UNCOMPRESSED_FORMAT = 1;
+    private static final long HASH_OUTPUT_HALF_RANGE = Long.MAX_VALUE;
+    private static final int SIZE_OF_KHYPERLOGLOG = ClassLayout.parseClass(KHyperLogLog.class).instanceSize();
+    private static final int SIZE_OF_RBTREEMAP = ClassLayout.parseClass(Long2ObjectRBTreeMap.class).instanceSize();
+
+    private final Long2ObjectSortedMap<HyperLogLog> minhash;
+    private final int maxSize;
+    private final int hllBuckets;
+
+    public KHyperLogLog()
+    {
+        this(DEFAULT_MAX_SIZE, DEFAULT_HLL_BUCKETS, new Long2ObjectRBTreeMap<>());
+    }
+
+    public KHyperLogLog(int maxSize, int hllBuckets)
+    {
+        this(maxSize, hllBuckets, new Long2ObjectRBTreeMap<>());
+    }
+
+    public KHyperLogLog(int maxSize, int hllBuckets, Long2ObjectSortedMap<HyperLogLog> minhash)
+    {
+        this.maxSize = maxSize;
+        this.hllBuckets = hllBuckets;
+        this.minhash = requireNonNull(minhash, "minhash is null");
+    }
+
+    public static KHyperLogLog newInstance(Slice serialized)
+    {
+        requireNonNull(serialized, "serialized is null");
+        SliceInput input = serialized.getInput();
+        checkArgument(input.readByte() == UNCOMPRESSED_FORMAT, "Unexpected version");
+
+        int maxSize = input.readInt();
+        int hllBuckets = input.readInt();
+        int minhashSize = input.readInt();
+
+        Long2ObjectRBTreeMap<HyperLogLog> minhash = new Long2ObjectRBTreeMap<>();
+        // The values are stored after the keys
+        SliceInput valuesInput = serialized.getInput();
+        valuesInput.setPosition(input.position() + minhashSize * SIZE_OF_LONG);
+
+        int hllLength;
+        Slice serializedHll;
+        for (int i = 0; i < minhashSize; i++) {
+            hllLength = valuesInput.readInt();
+            serializedHll = Slices.allocate(hllLength);
+            valuesInput.readBytes(serializedHll, hllLength);
+            minhash.put(input.readLong(), HyperLogLog.newInstance(serializedHll));
+        }
+
+        return new KHyperLogLog(maxSize, hllBuckets, minhash);
+    }
+
+    public static long exactIntersectionCardinality(KHyperLogLog a, KHyperLogLog b)
+    {
+        checkState(a.isExact(), "exact intersection cannot operate on approximate sets");
+        checkArgument(b.isExact(), "exact intersection cannot operate on approximate sets");
+
+        return Sets.intersection(a.minhash.keySet(), b.minhash.keySet()).size();
+    }
+
+    public static double jaccardIndex(KHyperLogLog a, KHyperLogLog b)
+    {
+        int sizeOfSmallerSet = Math.min(a.minhash.size(), b.minhash.size());
+        LongSortedSet minUnion = new LongRBTreeSet(a.minhash.keySet());
+        minUnion.addAll(b.minhash.keySet());
+
+        int intersection = 0;
+        int i = 0;
+        for (long key : minUnion) {
```

Comment:
This will end up promoting the primitive longs into java.lang.Long, let's find a way that avoids this.

Reply:
Changed implementation to follow mergeWith style.